### ISSUE TITLE
Pseudoinverse

### ISSUE THREAD

**Issue by el3ment** [pytorch/pytorch#6187]
PyTorch has `torch.inverse`, but given the nature of noisy data-driven problems it would be useful to have a `torch.pinverse` option which is calculated using the SVD.

As a side note -- it would also be great if these two functions (`.inverse` and the proposed `.pinverse`) were more optimized for a batch operation. As in, passing a tensor of shape (32, 10, 10) would perform 32 instances of inverse on 10x10 matricies, or passing in a tensor of shape (32, 8, 4, 4) would perform 32 * 8 inverses.

**Comment by SsnL** [comment]
Thanks for your request. We are currently working on other tasks, but we are putting this on our list. :)

**Comment by Naman-ntc** [comment]
@SsnL I would like to try this out! Is it possible that I work over this? I am up for reading!

**Comment by SsnL** [comment]
@Naman-ntc Go for it. Shouldn't be too much code using svd. Notice that you should also have gradcheck tests for this one.

**Comment by Naman-ntc** [comment]
Ohh, just point me to maybe relevant code/files maybe? (still unable to find location of various functions :P )

**Comment by SsnL** [comment]
@Naman-ntc Implementing this as a native function residing in `LinearAlgebra.cpp`. There should be plenty of similar examples.

**Comment by el3ment** [comment]
As a side note - I have implemented this a few different ways, and there are some numerical properties that need to be kept in mind. Using the SVD can create nan problems for backprop when the singular values are not unique. You can use the left and right pseudoinverses (using an if statement based on the shape) which involve the true inverse of a square `(A.T.dot(A)).inv()` or `(A.dot(A.T)).inv()` which seems to be more stable for backprop. Becuase those matricies are symmetric, you can also take advantage of other inverse algorithms like `torch.potri` but in practice, this doesn't seem to work as well when doing backprop

**Comment by el3ment** [comment]
A good test case will be looking at something like:

```
A = torch.eye(3)
B = torch.pinv(A)
L = torch.mean(B)
L.backward()
print(A.grad)
```
